File: src/main/java/com/hemanth/automation/pages/ProductsPage.java
package com.hemanth.automation.pages;

import org.openqa.selenium.By;

public class ProductsPage extends BasePage {

    private final By allProductsHeader = By.xpath("//h2[contains(.,'All Products')]");
    private final By firstProductName = By.xpath("(//div[@class='productinfo text-center']/p)[1]");
    private final By firstAddToCartButton =
        By.xpath("(//div[@class='productinfo text-center']//a[contains(@class,'add-to-cart')])[1]");
    private final By viewCartModalLink = By.cssSelector(".modal-content a[href='/view_cart']");

    private final By womenDressCategoryLink = By.cssSelector("a[href='/category_products/1']");
    private final By menTshirtsCategoryLink = By.cssSelector("a[href='/category_products/3']");
    private final By continueShoppingButton = By.xpath("//button[text()='Continue Shopping']");

    public boolean isAllProductsHeaderDisplayed() {

        return isDisplayed(allProductsHeader);
    }

    public String getFirstProductName() {
        return getText(firstProductName);
    }

    public void addFirstProductToCart() {
        dismissAdOverlayIfPresent();
        jsClick(firstAddToCartButton);
    }

    public void goToCartFromModal() {
        dismissAdOverlayIfPresent();
        click(viewCartModalLink);
    }

    public void selectWomenDressCategory() {
        dismissAdOverlayIfPresent();
        jsClick(womenDressCategoryLink);
    }

    public void selectMenTshirtsCategory() {
        dismissAdOverlayIfPresent();
        jsClick(menTshirtsCategoryLink);
    }

    public boolean isCategoryPageDisplayed(String categoryTitle) {
        By categoryHeader = By.xpath("//h2[contains(.,'" + categoryTitle + "')]");
        return isDisplayed(categoryHeader);
    }

    public String getProductName(int index) {
        By productName = By.xpath("(//div[@class='productinfo text-center']/p)[" + index + "]");
        return getText(productName);
    }

    public void addProductToCart(int index) {
        dismissAdOverlayIfPresent();
        By addToCartButton = By.xpath(
            "(//div[@class='productinfo text-center']//a[contains(@class,'add-to-cart')])[" + index + "]");
        jsClick(addToCartButton);
    }

    public void clickContinueShopping() {
        click(continueShoppingButton);
    }
}

File: playwright-ts/pages/ProductsPage.ts
import { expect, Locator, Page } from "@playwright/test";
import { clickWithAdGuard, dismissAdOverlay } from "../utils/adOverlay";

export class ProductsPage {
  readonly page: Page;
  readonly allProductsHeader: Locator;
  readonly firstProductName: Locator;
  readonly firstAddToCartButton: Locator;
  readonly viewCartModalLink: Locator;
  readonly womenPanelToggle: Locator;
  readonly womenDressCategoryLink: Locator;
  readonly menPanelToggle: Locator;
  readonly menTshirtsCategoryLink: Locator;
  readonly continueShoppingButton: Locator;

  constructor(page: Page) {
    this.page = page;
    this.allProductsHeader = page.locator(
      "xpath=//h2[contains(.,'All Products')]",
    );
    this.firstProductName = page.locator(".productinfo.text-center p").first();
    this.firstAddToCartButton = page
      .locator(".productinfo.text-center a.add-to-cart")
      .first();
    this.viewCartModalLink = page.locator(
      ".modal-content a[href='/view_cart']",
    );
    this.womenPanelToggle = page.locator("a[href='#Women']");
    this.womenDressCategoryLink = page.locator(
      "a[href='/category_products/1']",
    );
    this.menPanelToggle = page.locator("a[href='#Men']");
    this.menTshirtsCategoryLink = page.locator(
      "a[href='/category_products/3']",
    );
    this.continueShoppingButton = page.getByRole("button", {
      name: "Continue Shopping",
    });
  }

  async verifyAllProductsHeaderVisible() {
    await dismissAdOverlay(this.page);
    await expect(this.allProductsHeader).toBeVisible();
  }

  async getFirstProductName(): Promise<string> {
    return (await this.firstProductName.textContent()) ?? "";
  }

  async addFirstProductToCart() {
    await clickWithAdGuard(this.firstAddToCartButton);
  }

  async goToCartFromModal() {
    await clickWithAdGuard(this.viewCartModalLink);
  }

  async selectWomenDressCategory() {
    await clickWithAdGuard(this.womenPanelToggle);
    await clickWithAdGuard(this.womenDressCategoryLink);
  }

  async selectMenTshirtsCategory() {
    await clickWithAdGuard(this.menPanelToggle);
    await clickWithAdGuard(this.menTshirtsCategoryLink);
  }

  async verifyCategoryPageDisplayed(categoryTitle: string) {
    await expect(
      this.page.locator(`xpath=//h2[contains(.,'${categoryTitle}')]`),
    ).toBeVisible();
  }

  async getProductName(index: number): Promise<string> {
    return (
      (await this.page
        .locator(".productinfo.text-center p")
        .nth(index - 1)
        .textContent()) ?? ""
    );
  }

  async addProductToCart(index: number) {
    await clickWithAdGuard(
      this.page.locator(".productinfo.text-center a.add-to-cart").nth(index - 1),
    );
  }

  async clickContinueShopping() {
    await this.continueShoppingButton.click();
  }
}
